File: 0103_Binary_Tree_Zigzag_Level_Order_Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        LinkedList<TreeNode> queue = new LinkedList<>();
        ArrayList<List<Integer>> result = new ArrayList<>();
        boolean isReversed = false;
        
        if(root ==  null) return result;
        
        queue.add(root);
        queue.add(null); // null signifies the end of a level
        
        LinkedList<Integer> level = new LinkedList<>();
        
        while(queue.size() > 1) {
            TreeNode front = queue.remove(0);
            if(front == null) { // level is finished
                result.add(level);
                level = new LinkedList();
                isReversed = !isReversed;
                queue.add(null);
            }  else {
                if(front.left != null) queue.add(front.left);
                if(front.right != null) queue.add(front.right);
                if(isReversed) level.add(0, front.val);
                else level.add(front.val);
            }
        }
        result.add(level);
        
        return result;
    }
}

File: 0103_Binary_Tree_Zigzag_Level_Order_Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levels;
    
    void inorderTraverse(TreeNode* node, int breadth = 0) {
        if(node == nullptr) return;
        while(levels.size() <= breadth) levels.push_back({}); // fill levels with blank arrays, as needed
        
        inorderTraverse(node->left, breadth + 1);
        levels[breadth].push_back(node->val);
        inorderTraverse(node->right, breadth + 1);
    }
    
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        inorderTraverse(root); // fill %levels% with nodes, left-to-right
        
        for(int i = 1; i < levels.size(); i += 2) { // reverse every odd row
            reverse(levels[i].begin(), levels[i].end());
        }
        
        return levels;
    }
};
